### PR TITLE
Add config editing via the CLI for storage nodes

### PR DESCRIPTION
**Pull request by wthorp**
"When a SNO runs setup they should have the option to edit the config directly"
In addition, we'd really like to have comments in the configs so that people know what the config items do.

### CODE REVIEW

**Comment by wthorp** [comment]
We previously generated the YAML files with Viper (https://github.com/spf13/viper) by scraping for flags declared in the code, which in turn serializes the YAML using Go-yaml (https://github.com/go-yaml/yaml).

Viper doesn't support comments, and Go-yaml doesn't support writing comments either.  There are some forks and PRs out there that do.

However, it sounds like there's a Go-yaml V3 in the works which sort of throws a wrench in things.
Soon, go-yaml should have native-ish comment support.  Currently it does not.

---
"thanks for your contributions, but please note that I've been working on a new v3 version which I hope to release this year still, which will render these changes incompatible. There I've also been working on initial support for comments, but given what I see in your comments above it will likely not be as complete as your PR seems to be. On the bright side, the main change in v3 is the exposure of an intermediate state, which is much more appropriate for preserving comments (and style!) than what we currently have in v2. I'm currently in the final stages of finishing that (testing and last fixes mainly) before exposing an initial version for everyone to evaluate.

Might be best to pause a bit this work, and re-engage once v3 is ready for testing in a few days."
---

https://github.com/go-yaml/yaml/issues/132
https://github.com/go-yaml/yaml/pull/419

**Comment by mniewrzal** [comment]
I have generated captplanet config file with this PR and I have two initial comments. First, structure is flat now and I think its less readable. Second, boolean flags are saved as string e.g. `level: debug` vs `log.level: "debug"`

**Comment by wthorp** [comment]
I absolutely knew the structure would be flat; what is more readable is debatable.  This was _easy_.  The whitespace sensitivity of YAML actually causes problems with a lot of people in the real world, so I think the merits on either side are debatable.

You're definitely right that I should make the value serialization more robust.  I will try to pull in whatever logic go-yaml uses to serialize values; as my logic would clearly fail to escape certain things correctly.

**Comment by mniewrzal** [comment]
Definitely flat vs hierarchical is debatable :)

File: internal/fpath/editor.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	git, err := exec.LookPath("git")
+	if err == nil {
+		out, err := exec.Command(git, "config", "core.editor").Output()
+		cmd := strings.TrimSpace(string(out))
+		if err == nil && len(cmd) > 0 {
+			return cmd
+		}
+	}
+	//heck, just try a bunch of options
+	for _, exe := range [...]string{"nvim", "vim", "emacs", "nano", "pico"} {
```
[the remark this answers is not in the document]
Comment (wthorp) [reply internal/fpath/editor.go]: added vi

**Comment by wthorp** [comment]
I looked at how to make the YAML serialization sexier, but it would be 300+ lines of code with little real value.  I did opt to remove quotes from non-string YAML, we'll have to see if that causes any issues, but I don't think so.

File: cmd/storagenode/main.go [hunk 2]
```diff
@@ -43,6 +43,11 @@ var (
 		RunE:        cmdSetup,
 		Annotations: map[string]string{"type": "setup"},
 	}
+	configCmd = &cobra.Command{
+		Use:   "config",
+		Short: "Edit config files",
+		RunE:  cmdConfig,
```
Comment (mniewrzal) [inline cmd/storagenode/main.go]: You need to add `Annotations: map[string]string{"type": "setup"},` here to make `setup` working inside this command.

**Pull request merged**